## Project Overview
- This project focuses on predicting used car selling prices using machine learning.
- The goal is to build a model that helps buyers, sellers, dealerships, and online platforms make informed pricing decisions,
  reduce guesswork, and improve trust in transactions.


## Business Problem:

- Used car pricing is highly variable, depending on factors such as age, mileage, fuel type, and ownership history.
- Predicting accurate prices benefits both customers and businesses by ensuring fair deals and efficient decision-making.

### Connecting to databases

- The first step is to connect to postgresql for car sales data
- Here are the libraries needed for this connection.
- psycopg2-binary: A PostgreSQL adapter for Python to communicate with the database.
- sqlalchemy: An SQL toolkit and Object-Relational Mapping (ORM) library. Used cmd to get it. {pip install psycopg2-binary sqlalchemy pandas}
- pandas: To easily manipulate and load data into data frames.

### Step 1: Import Libraries and Connect to PostgreSQL
- In your Jupyter Notebook, you'll need to import the required libraries and set up the connection to your PostgreSQL database.

In [1]:
#Import pandas library for data manipulation and analyze the data
import pandas as pd
import numpy as np
# First, install the required PostgreSQL driver
!pip install psycopg2-binary

# Then import create_engine and establish the connection
from sqlalchemy import create_engine

import warnings
warnings.filterwarnings('ignore')

## 2. Create a Database Connection
- You can use SQLAlchemy to create a connection to your PostgreSQL database.
- username: Your PostgreSQL username (e.g., postgres).yes its postgres
- password: 2025.
- localhost: The address of the PostgreSQL server (if it's on your local machine, it will be localhost).
- 5432: Default port for PostgreSQL (can be different if customized).
- Car_Sales: The name of the database you created earlier.

In [2]:
# Replace 'username', 'password', 'localhost', 'car_sales' with your actual PostgreSQL credentials
engine = create_engine('postgresql://postgres:2025@localhost:5432/Car_Sales')


## 3. Query the Database
- Now that you’ve established the connection, you can query the database using pandas to load data into a DataFrame.

In [3]:
# SQL query to fetch all rows from the car_sales_data table
query = "SELECT * FROM car_sales_data"

# Load the data into a pandas DataFrame
df = pd.read_sql(query, engine)

data = df.copy()

## DATABASES FAILED TO CREATE CONNECTION NOW OPTED TO USE DRIVE

In [4]:
# Load the data from the CSV file into a pandas DataFrame
#df = pd.read_excel('/content/car data.xlsx')

# Create a copy of the DataFrame
#data = df.copy()

# Display the first few rows of the DataFrame to verify
#display(data.head())

In [5]:
data.head()

,car_name,year,present_price,kms_driven,fuel_type,seller_type,transmission,owner,selling_price
0,Hero Passion X pro,2016,12417.55,33217,Petrol,Individual,Manual,0,11717.50
1,xcent,2017,10633.13,12905,Petrol,Dealer,Manual,0,9931.75
2,verna,2015,10881.40,62053,Diesel,Dealer,Manual,0,10180.25
3,ertiga,2016,11107.79,43897,Diesel,Dealer,Manual,0,10404.75
4,jazz,2016,10316.40,4108,Petrol,Dealer,Manual,0,9614.00


In [6]:
data.tail()

,car_name,year,present_price,kms_driven,fuel_type,seller_type,transmission,owner,selling_price
99995,ertiga,2015,11994.71,27787,Petrol,Dealer,Manual,0,11293.10
99996,corolla altis,2017,11725.64,10207,Petrol,Dealer,Manual,0,11024.00
99997,verna,2013,9601.40,46392,Diesel,Dealer,Manual,0,8898.15
99998,KTM RC200,2017,10846.78,4645,Petrol,Individual,Manual,0,10146.65
99999,i10,2012,9464.60,37035,Petrol,Dealer,Manual,0,8763.10


### Understand the shape of the dataset

In [7]:
data.shape

(100000, 9)

## check the datatypes of the columns for the dataset

In [8]:
data.dtypes

car_name          object
year               int64
present_price    float64
kms_driven         int64
fuel_type         object
seller_type       object
transmission      object
owner              int64
selling_price    float64
dtype: object

**Comment on Data Types:**

The output shows the data types of each column in the DataFrame:
- `object`: Indicates categorical or string data (e.g., `car_name`, `fuel_type`). These will need encoding for machine learning models.
- `int64`: Integer numerical data (e.g., `year`, `kms_driven`, `owner`).
- `float64`: Floating-point numerical data (e.g., `present_price`, `selling_price`).

These data types are appropriate for the information they represent, but the 'object' types will require conversion to numerical formats before modeling.

## Checking the statistical summary of the data

In [9]:
data.describe().T  # summary stats like mean, std, min, max (transposed for readability)

,count,mean,std,min,25%,50%,75%,max
year,100000.0,2014.059300,2.592811,2003.00,2013.0000,2015.000,2016.0000,2018.00
present_price,100000.0,10454.815638,1497.175989,3405.28,9406.0075,10628.705,11639.7825,12731.23
kms_driven,100000.0,35544.731010,35908.164469,606.00,15760.0000,30595.500,46476.0000,502479.00
owner,100000.0,0.036080,0.213913,0.00,0.0000,0.000,0.0000,3.00
selling_price,100000.0,9752.087633,1498.468367,2703.35,8703.2500,9926.700,10938.1200,12028.00


**Comment on Statistical Summary:**

The transposed statistical summary provides key insights into the numerical features:
- **Count:** All numerical columns have 100,000 non-null entries, indicating no missing values in these columns.
- **Mean and Median (50%):** Comparing the mean and median helps identify skewness. For `year`, `present_price`, `owner`, and `selling_price`, the mean and median are relatively close, suggesting a roughly symmetrical distribution. `kms_driven`, however, has a mean (35544) higher than its median (30595), which could indicate a right skew or the presence of outliers with very high mileage.
- **Standard Deviation (std):** This shows the spread of the data. `kms_driven` has a much larger standard deviation (35908) compared to its mean, confirming a wide spread of values. `present_price` and `selling_price` have similar standard deviations (around 1497 and 1498 respectively), suggesting a similar level of variability in their prices.
- **Min and Max:** These values highlight the range of the data. The `kms_driven` column has a very high maximum value (502479), which is significantly larger than the 75th percentile (46476). This reinforces the possibility of outliers in mileage. The `owner` column shows values from 0 to 3, indicating a small number of distinct owner categories.

Overall, the summary suggests the data is generally well-behaved in numerical columns except for `kms_driven`, which might require outlier handling or transformation due to its potential skewness and large maximum value.

## Checking for duplicate values


In [10]:
# checking for duplicate values
data.duplicated().sum()

np.int64(6935)

## Check for missing values

In [11]:
# checking for mi((ssing values in the data
data.isnull().sum()#Complete the code to compute the number of missing values.

car_name         0
year             0
present_price    0
kms_driven       0
fuel_type        0
seller_type      0
transmission     0
owner            0
selling_price    0
dtype: int64

## Exploratory Data Analysis (EDA)

### Univariate Analysis


## Data Overview / Exploratory Data Analysis

Exploratory Data Analysis (EDA) is an approach to analyzing datasets to summarize their main characteristics, often with visual methods. EDA is a critical first step in the data analysis process, where the goal is to gain insights into the structure, relationships, and patterns of the data before diving into more complex statistical analysis or modeling. The purpose of EDA is to:

- Understand the data's underlying structure.
- Identify potential issues such as missing values, outliers, and errors.
- Determine which variables are most relevant for the analysis.
- Visualize key relationships between variables.
- Test assumptions that are necessary for further analysis or modeling.

#### EDA is typically carried out using various techniques, such as:

- Descriptive Statistics: Summarizing data with metrics like mean, median, mode, standard deviation, and percentiles.
- Visualizations: Creating plots like histograms, box plots, scatter plots, and heatmaps to understand data distributions and relationships.
- Missing Data Analysis: Identifying missing values and deciding how to handle them (e.g., imputation or deletion).
- Outlier Detection: Identifying extreme values that may skew results or indicate data entry errors.
- Correlation Analysis: Identifying relationships between variables to see if one influences the other.

In [12]:
# function to plot a boxplot and a histogram along the same scale.
import matplotlib.pyplot as plt
import seaborn as sns

def histogram_boxplot(data, feature, figsize=(12, 7), kde=False, bins=None):
    """
    Boxplot and histogram combined

    data: dataframe
    feature: dataframe column
    figsize: size of figure (default (12,7))
    kde: whether to the show density curve (default False)
    bins: number of bins for histogram (default None)
    """
    f2, (ax_box2, ax_hist2) = plt.subplots(
        nrows=2,  # Number of rows of the subplot grid= 2
        sharex=True,  # x-axis will be shared among all subplots
        gridspec_kw={"height_ratios": (0.25, 0.75)},
        figsize=figsize,
    )  # creating the 2 subplots
    sns.boxplot(
        data=data, x=feature, ax=ax_box2, showmeans=True, color="violet"
    )  # boxplot will be created and a star will indicate the mean value of the column
    sns.histplot(
        data=data, x=feature, kde=kde, ax=ax_hist2, bins=bins, palette="winter"
    ) if bins else sns.histplot(
        data=data, x=feature, kde=kde, ax=ax_hist2
    )  # For histogram
    ax_hist2.axvline(
        data[feature].mean(), color="green", linestyle="--"
    )  # Add mean to the histogram
    ax_hist2.axvline(
        data[feature].median(), color="black", linestyle="-"
    )  # Add median to the histogram


A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.3.3 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\LENOVO\anaconda3\Lib\site-packages\ipykernel_launcher.py", line 17, in <module>
    app.launch_new_instance()
  File "C:\Users\LENOVO\anaconda3\Lib\site-packages\traitlets\config\application.py", line 1075, in launch_instance
    app.start()
  File "C:\Users\LENOVO\anaconda3\Lib\site-packages\ipykernel\kernelapp.py", line 701, in start
    self.io_loop.start()
  File "C:\Users\LENOVO\anaconda3\Lib\site-

ImportError: 
A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.3.3 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.




A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.3.3 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\LENOVO\anaconda3\Lib\site-packages\ipykernel_launcher.py", line 17, in <module>
    app.launch_new_instance()
  File "C:\Users\LENOVO\anaconda3\Lib\site-packages\traitlets\config\application.py", line 1075, in launch_instance
    app.start()
  File "C:\Users\LENOVO\anaconda3\Lib\site-packages\ipykernel\kernelapp.py", line 701, in start
    self.io_loop.start()
  File "C:\Users\LENOVO\anaconda3\Lib\site-

ImportError: 
A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.3.3 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.




A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.3.3 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\LENOVO\anaconda3\Lib\site-packages\ipykernel_launcher.py", line 17, in <module>
    app.launch_new_instance()
  File "C:\Users\LENOVO\anaconda3\Lib\site-packages\traitlets\config\application.py", line 1075, in launch_instance
    app.start()
  File "C:\Users\LENOVO\anaconda3\Lib\site-packages\ipykernel\kernelapp.py", line 701, in start
    self.io_loop.start()
  File "C:\Users\LENOVO\anaconda3\Lib\site-

ImportError: 
A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.3.3 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.



ValueError: numpy.dtype size changed, may indicate binary incompatibility. Expected 96 from C header, got 88 from PyObject

**Why Univariate Analysis is Necessary:**

Univariate analysis is a fundamental step in Exploratory Data Analysis (EDA). It's necessary for several reasons:

- **Understanding Individual Features:** It allows you to examine each variable in isolation to understand its basic characteristics, distribution, and range.
- **Identifying Data Quality Issues:** You can easily spot missing values, outliers, and potential data entry errors by looking at the distribution of a single variable.
- **Choosing Appropriate Analysis Methods:** The distribution of a variable (e.g., normal, skewed, categorical) helps you decide which statistical tests or machine learning models are appropriate to use later on.
- **Gaining Initial Insights:** It provides a foundational understanding of your data before you start looking at relationships between multiple variables.

In short, univariate analysis is the crucial first step to getting acquainted with your dataset, ensuring data quality, and preparing for more complex analysis.

In [ ]:
histogram_boxplot(data, "Year")

**Comment on Year Distribution:**

The plots for the 'Year' column show the distribution of the manufacturing years of the cars in the dataset.
- The **boxplot** indicates the median year and the spread of the data, as well as any potential outliers.
- The **histogram** shows the frequency of cars manufactured in each year.

Looking at these plots, you can observe the most common manufacturing years and how the data is distributed across different years. This helps understand the age profile of the cars in the dataset.

In [ ]:
histogram_boxplot(data, "Present_Price")

**Comment on Present_Price Distribution:**

The plots for 'Present_Price' visualize the distribution of the current prices of the cars in the dataset.
- The **boxplot** provides a summary of the price distribution, showing the median, quartiles, and any detected outliers.
- The **histogram** illustrates the frequency of cars at different price levels, allowing you to see which price ranges are most common.

Examining these plots helps you understand the typical market price for these used cars and identify if there are any cars with unusually high or low present prices compared to the rest of the dataset.

In [ ]:
histogram_boxplot(data, "Kms_Driven")

**Comment on Kms_Driven Distribution:**

The plots for 'Kms_Driven' display the distribution of the mileage of the cars in the dataset.
- The **boxplot** provides a summary of the mileage distribution, showing the median, quartiles, and highlighting potential outliers with very high mileage, consistent with what was observed in the statistical summary.
- The **histogram** shows the frequency of cars within different mileage ranges. The distribution appears to be skewed to the right, with a long tail towards higher mileage values.

These plots confirm the presence of cars with significantly higher mileage compared to the majority of the dataset, which might require further investigation or handling during feature engineering.

In [ ]:
histogram_boxplot(data, "Owner")

**Comment on Owner Distribution:**

The plots for the 'Owner' column display the distribution of the number of previous owners for the cars in the dataset.
- The **boxplot** and **histogram** show the frequency of cars with 0, 1, 2, or 3 previous owners.

These plots reveal the distribution of car ownership history, showing which number of previous owners is most common in the dataset. This information can be useful for understanding how ownership history might relate to other features or the selling price.

In [ ]:
histogram_boxplot(data, "Selling_Price")

**Comment on Selling_Price Distribution:**

The plots for 'Selling_Price' display the distribution of the final selling prices of the cars in the dataset.
- The **boxplot** provides a summary of the selling price distribution, showing the median, quartiles, and potential outliers.
- The **histogram** shows the frequency of cars within different selling price ranges.

Examining these plots helps you understand the typical selling prices in your dataset and the overall distribution of your target variable.

In [ ]:
# function to create labeled barplots


def labeled_barplot(data, feature, perc=False, n=None):
    """
    Barplot with percentage at the top

    data: dataframe
    feature: dataframe column
    perc: whether to display percentages instead of count (default is False)
    n: displays the top n category levels (default is None, i.e., display all levels)
    """

    total = len(data[feature])  # length of the column
    count = data[feature].nunique()
    if n is None:
        plt.figure(figsize=(count + 1, 5))
    else:
        plt.figure(figsize=(n + 1, 5))

    plt.xticks(rotation=90, fontsize=15)
    ax = sns.countplot(
        data=data,
        x=feature,
        palette="Paired",
        order=data[feature].value_counts().index[:n].sort_values(),
    )

    for p in ax.patches:
        if perc == True:
            label = "{:.1f}%".format(
                100 * p.get_height() / total
            )  # percentage of each class of the category
        else:
            label = p.get_height()  # count of each level of the category

        x = p.get_x() + p.get_width() / 2  # width of the plot
        y = p.get_height()  # height of the plot

        ax.annotate(
            label,
            (x, y),
            ha="center",
            va="center",
            size=12,
            xytext=(0, 5),
            textcoords="offset points",
        )  # annotate the percentage

    plt.show()  # show the plot

**Why Labeled Barplots are Necessary:**

While basic bar plots (like those created by `seaborn.countplot` directly) show the relative frequencies of categories, adding labels directly to the bars makes the visualization much more informative.

- **Easy Interpretation:** You can instantly see the exact count or percentage for each category without having to guess from the axis scale.
- **Precise Comparisons:** It's easier to compare the exact values of different categories.
- **Highlights Distribution:** Clearly shows which categories are most or least frequent.

The `labeled_barplot` function automates the process of adding these labels, saving you from writing the annotation code manually for each plot. It's especially useful when you have several categorical features to visualize and want a consistent, clear presentation.

In [ ]:
labeled_barplot(data, "Fuel_Type", perc=True)

**Comment on Fuel_Type Distribution:**

The bar plot for 'Fuel_Type' shows the distribution of different fuel types in the dataset, with percentages labeled on each bar. This visualization clearly indicates:
- The most prevalent fuel type(s).
- The proportion of cars for each fuel type (Petrol, Diesel, CNG).

This helps understand the fuel type composition of the used cars in your dataset and is important for feature engineering, as fuel type can significantly impact selling price and might require encoding.

In [ ]:
labeled_barplot(data, "Seller_Type", perc=True)

**Comment on Seller_Type Distribution:**

The bar plot for 'Seller_Type' displays the distribution of different seller types in the dataset, with percentages labeled on each bar. This visualization clearly shows:
- The proportion of cars sold by Dealers versus Individuals.

Understanding the distribution of seller types is important as it can influence the selling price and may require encoding for machine learning models.

In [ ]:
labeled_barplot(data, "Transmission", perc=True)

**Comment on Transmission Distribution:**

The bar plot for 'Transmission' displays the distribution of different transmission types in the dataset, with percentages labeled on each bar. This visualization clearly shows:
- The proportion of cars with Manual versus Automatic transmission.

Understanding the distribution of transmission types is important as it can influence the selling price and may require encoding for machine learning models.

## Bivariate analysis



**Why Bivariate Analysis is Important at this Stage:**

After understanding each variable individually through univariate analysis, **bivariate analysis** is the next crucial step. It involves examining the relationship between *two* variables at a time. It's important at this stage for several reasons:

- **Identifying Relationships:** It helps uncover how two variables relate to each other. For example, how does 'Kms_Driven' relate to 'Selling_Price', or how does 'Year' relate to 'Present_Price'?
- **Understanding Feature Interactions:** Sometimes, the interaction between two features is more informative than the features in isolation. Bivariate analysis can highlight these interactions.
- **Selecting Features for Modeling:** By understanding the relationships, you can identify which features have a strong correlation with your target variable ('Selling_Price') and which features might be redundant or less important. The correlation matrix you just generated is a key tool for this with numerical features.
- **Detecting Patterns and Trends:** Visualizations in bivariate analysis (like scatter plots for numerical-numerical relationships, or box plots/violin plots for numerical-categorical relationships) can reveal patterns, trends, and differences between groups.
- **Informing Feature Engineering:** Understanding relationships can guide you in creating new features or transforming existing ones. For example, if there's a strong non-linear relationship, a transformation might be needed.

In essence, bivariate analysis moves beyond describing individual features to understanding the connections and dependencies within your data, which is vital for building an effective predictive model.

In [ ]:
# Calculate the correlation matrix
correlation_matrix = data.corr(numeric_only=True)

# Visualize the correlation matrix using a heatmap
plt.figure(figsize=(12, 8))
sns.heatmap(correlation_matrix, annot=True, cmap='coolwarm', fmt=".2f")
plt.title("Correlation Matrix of Numerical Features")
plt.show()

**Comment on Correlation Matrix:**

The correlation matrix provides insights into the linear relationships between numerical features:

- **'Selling_Price' and 'Present_Price'**: There is a very strong positive correlation (0.999996, almost 1.0) between 'Selling_Price' and 'Present_Price'. This indicates that the current price of the car is highly predictive of its selling price, which makes intuitive sense.
- **'Selling_Price' and 'Year'**: There is a strong positive correlation (0.783294) between 'Selling_Price' and 'Year'. This suggests that newer cars tend to have higher selling prices.
- **'Selling_Price' and 'Kms_Driven'**: There is a moderate negative correlation (-0.388601) between 'Selling_Price' and 'Kms_Driven'. This indicates that cars with higher mileage tend to have lower selling prices.
- **'Present_Price' and 'Year'**: There is a strong positive correlation (0.782933) between 'Present_Price' and 'Year', similar to the relationship between 'Selling_Price' and 'Year'. This is expected as the current price of a car is related to its age.
- **Other Correlations**: The correlations between 'Owner' and other variables are relatively weak (close to 0), suggesting that the number of previous owners has a less significant *linear* relationship with the other numerical features compared to 'Year', 'Present_Price', and 'Kms_Driven'. 'Kms_Driven' has a moderate negative correlation with 'Year' (-0.529906), which is also expected as older cars are likely to have been driven more.

Overall, the correlation matrix confirms expected relationships and highlights 'Present_Price', 'Year', and 'Kms_Driven' as potentially important features for predicting 'Selling_Price'. The very high correlation between 'Selling_Price' and 'Present_Price' suggests they are very closely related, which is something to consider during modeling (e.g., to avoid multicollinearity if both are used as predictors).

In [ ]:
!pip install datasist

In [ ]:
import datasist as ds

## We perform visualization with the library datasist

## Feature Engineering

Feature Engineering is the process of transforming raw data into meaningful features that improve the performance of machine learning models.
It plays a crucial role in preparing data for modeling and ensures that the model can learn from the data effectively.

#### Key Aspects of Feature Engineering:

- Feature Selection: Choosing relevant features and removing redundant or irrelevant ones.
- Feature Creation: Generating new features that provide more insight into the data.
- Feature Transformation: Applying techniques like normalization, scaling, or encoding categorical data.
- Handling Missing Data: Imputing or removing missing values and creating new features to capture missingness.
- Categorical Encoding: Converting categorical variables into numerical form using techniques like one-hot encoding or label encoding.

#### Why is Feature Engineering Important?

- Improves Model Accuracy: Quality features help models learn better patterns and make more accurate predictions.
- Prevents Overfitting: Reduces irrelevant features, ensuring the model generalizes better to unseen data.
- Handles Real-World Data: It helps manage challenges like missing data, noisy data, and unstructured formats.
- Enhances Interpretability: Meaningful features make it easier to understand how the model is making decisions.

In [ ]:
# create a feature that calculate how old is the car
data["Year"] = 2025 - data["Year"]

**Why Transforming 'Year' to Car Age is Necessary:**

Transforming the 'Year' column into 'Car Age' is a common and often necessary feature engineering step in predicting car prices for a few key reasons:

- **Direct Relationship with Value:** The age of a car (how many years it's been since manufacturing) usually has a more direct and often more linear relationship with its depreciation and selling price than the manufacturing year itself. An older car is generally worth less than a newer one, regardless of the specific manufacturing year.
- **Captures Depreciation:** Age is a primary driver of car depreciation. By using age as a feature, you are explicitly providing the model with information about how much the car has likely depreciated.
- **Avoids Temporal Dependencies:** Using the manufacturing year can sometimes introduce unnecessary temporal dependencies into the model. Using age makes the feature more about the intrinsic property of the car (how old it is) rather than a specific point in time.

In essence, 'Car Age' is often a more intuitive and powerful predictor of used car selling price than the manufacturing year, making this transformation a valuable part of preparing your data for modeling.

In [ ]:
data.head(5)

- We can see the year column with the age of the car .

## Select feature to use in this modelling

In [ ]:
data.columns

In [ ]:
# drop the car name columns
data = data [['Year', 'Present_Price', 'Kms_Driven', 'Fuel_Type',
       'Seller_Type', 'Transmission', 'Owner', 'Selling_Price']]

**Why 'Car_Name' Column Was Dropped:**

The 'Car_Name' column was dropped from the DataFrame for the purpose of this initial modeling phase. Here are some common reasons for dropping such a column:

- **High Cardinality:** The 'Car_Name' column likely contains many unique car names. Directly using such a feature in many machine learning models can be challenging due to high cardinality, requiring complex encoding or potentially leading to overfitting if not handled carefully.
- **Information Captured Elsewhere:** Much of the relevant information about a car's identity is often captured by other features like 'Year' (age), 'Present_Price', 'Fuel_Type', 'Seller_Type', and 'Transmission'. These features provide more structured and directly usable information for a regression model.
- **Simplification for Initial Model:** For a first iteration of a model, it's often simpler to focus on features that are more straightforward to include and interpret. While car name *can* be a valuable feature, it might require more advanced text processing or feature engineering techniques (like grouping by brand or model) which are being skipped in this step.

By dropping 'Car_Name', you are focusing on a set of features that are readily usable for modeling and have shown relevant correlations during the bivariate analysis.

In [ ]:
data.head()

#####  Perform an encoding techniques , ml cannot run with categorical data types , we need to convert cats to numerics

In [ ]:
data = pd.get_dummies(data, drop_first = True)

**Why Categorical Encoding is Necessary:**

Most machine learning algorithms are designed to work with numerical data. They rely on mathematical operations that cannot be performed on text-based or categorical values directly.

Therefore, **encoding categorical data into a numerical format is a necessary preprocessing step** before feeding the data into a machine learning model. Techniques like one-hot encoding (which you've applied here) create numerical representations that algorithms can understand and process.

Without encoding, models would not be able to learn patterns or relationships from your categorical features like 'Fuel_Type', 'Seller_Type', or 'Transmission', significantly hindering their ability to make accurate predictions.

In [ ]:
data.head()

- You can see the data is now encoded

In [ ]:
data.dtypes

In [ ]:
# Assuming df is your DataFrame
data['Fuel_Type_Diesel'] = data['Fuel_Type_Diesel'].astype(int)
data['Fuel_Type_Petrol'] = data['Fuel_Type_Petrol'].astype(int)
data['Seller_Type_Individual'] = data['Seller_Type_Individual'].astype(int)
data['Transmission_Manual'] = data['Transmission_Manual'].astype(int)

**Why Convert Boolean to Integer After One-Hot Encoding:**

After using `pd.get_dummies`, the new columns representing your categorical variables have a `bool` (boolean) data type (True/False). While some machine learning libraries can work directly with boolean types, converting them to integers (0 for False, 1 for True) is a common practice for several reasons:

- **Compatibility:** Many machine learning algorithms and libraries are primarily designed to work with numerical data (integers or floats). Explicitly converting booleans to integers ensures broader compatibility.
- **Consistency:** Keeping all your input features in numerical formats (integers or floats) provides data type consistency, which can simplify subsequent data processing and model building steps.
- **Interpretability (for some models):** For certain models, having 0s and 1s as explicit numerical values can sometimes be slightly more intuitive than True/False boolean values.

In summary, converting boolean dummy variables to integers (0s and 1s) is a standard practice to ensure data is in a fully numerical format, compatible with a wider range of machine learning tools and techniques.

In [ ]:
data.head()

## Do correlation to figure out signficant features

In [ ]:
import seaborn as sns
from matplotlib import pyplot as plt
plt.rcParams['figure.figsize']=[16,10]
sns.heatmap(data.corr(),annot = True)


The updated correlation matrix provides insights into the linear relationships between features, now including the 'Car Age' (transformed 'Year') and the one-hot encoded categorical variables.

- **'Selling_Price' vs. Other Features:**
    - **'Year' (Car Age):** Shows a strong *negative* correlation with 'Selling_Price' (around -0.78). This makes sense – older cars tend to have lower selling prices.
    - **'Present_Price':** Still has a very strong positive correlation with 'Selling_Price' (around 0.99), indicating that the current listed price is a very strong predictor of the final selling price.
    - **'Kms_Driven':** Has a moderate negative correlation with 'Selling_Price' (around -0.39), meaning higher mileage is associated with lower selling prices.
    - **'Seller_Type_Individual':** Shows a noticeable negative correlation with 'Selling_Price' (around -0.52). This suggests that cars sold by individuals tend to have lower selling prices compared to those sold by dealers.
    - **'Transmission_Manual':** Has a negative correlation with 'Selling_Price' (around -0.24). This indicates that manual transmission cars tend to have slightly lower selling prices than automatic ones in this dataset.
    - **'Fuel_Type_Diesel':** Shows a positive correlation with 'Selling_Price' (around 0.39). This suggests diesel cars might have higher selling prices on average compared to petrol or CNG.
    - **'Fuel_Type_Petrol':** Shows a negative correlation with 'Selling_Price' (around -0.36). This is the inverse of the diesel relationship, indicating petrol cars might have lower selling prices than diesel cars.
    - **'Owner':** Still shows a weak negative correlation with 'Selling_Price' (around -0.15), suggesting the number of previous owners has a relatively minor linear impact on the selling price compared to other factors.

- **Other Notable Correlations:**
    - 'Year' (Car Age) has a strong positive correlation with 'Kms_Driven' (around 0.53), which is expected as older cars are likely to have more mileage.
    - 'Fuel_Type_Diesel' and 'Fuel_Type_Petrol' have a strong negative correlation (-0.95), which is expected because they are dummy variables representing mutually exclusive categories (a car cannot be both diesel and petrol).

Overall, this matrix helps confirm which features have the strongest linear relationships with the selling price, guiding your feature selection for the model. The very high correlation between 'Selling_Price' and 'Present_Price' is significant and should be considered when building the model.

## Let's check the distribution of our target variable

In [ ]:
plt.figure(figsize=[8, 6])
sns.scatterplot(x=data.Fuel_Type_Diesel, y=data.Selling_Price)
plt.show()

**Comment on Fuel_Type_Diesel vs. Selling_Price Scatter Plot:**

This scatter plot visualizes the relationship between whether a car is diesel (`Fuel_Type_Diesel` = 1) or not (`Fuel_Type_Diesel` = 0) and its `Selling_Price`. You will observe two vertical clusters of points:

- A cluster at `x=0`: Represents the selling prices of non-diesel cars.
- A cluster at `x=1`: Represents the selling prices of diesel cars.

By comparing the distribution of points along the y-axis (Selling_Price) for these two clusters, you can visually assess if there is a difference in selling prices between diesel and non-diesel cars. This plot supports the positive correlation observed in the heatmap, suggesting that diesel cars tend to have different selling prices compared to other fuel types in this dataset.

In [ ]:
plt.figure(figsize=[8, 6])
sns.scatterplot(x=data.Fuel_Type_Petrol, y=data.Selling_Price)
plt.show()

**Comment on Fuel_Type_Petrol vs. Selling_Price Scatter Plot:**

This scatter plot visualizes the relationship between whether a car is petrol (`Fuel_Type_Petrol` = 1) or not (`Fuel_Type_Petrol` = 0) and its `Selling_Price`. Similar to the diesel plot, you will see two vertical clusters of points:

- A cluster at `x=0`: Represents the selling prices of non-petrol cars (likely Diesel and CNG).
- A cluster at `x=1`: Represents the selling prices of petrol cars.

By comparing the distribution of points along the y-axis (Selling_Price) for these two clusters, you can visually assess if there is a difference in selling prices between petrol and non-petrol cars. This plot supports the negative correlation observed in the heatmap for `Fuel_Type_Petrol`, suggesting that petrol cars tend to have different selling prices compared to other fuel types in this dataset.

In [ ]:
plt.figure(figsize=[8, 6])
sns.scatterplot(x=data.Kms_Driven, y=data.Selling_Price)
plt.show()

**Comment on Kms_Driven vs. Selling_Price Scatter Plot:**

This scatter plot visualizes the relationship between the `Kms_Driven` (mileage) and the `Selling_Price`.

- You can generally observe a trend where as `Kms_Driven` increases (moving to the right on the x-axis), the `Selling_Price` tends to decrease (moving downwards on the y-axis). This reflects the negative correlation seen in the heatmap – cars with higher mileage are generally sold for less.
- You might also notice that the points are more spread out at lower mileages, and potentially cluster or show a clearer downward trend as mileage increases.
- Look for any potential outliers, which would appear as points far away from the main cluster or trend.

This plot provides a visual confirmation of how mileage impacts selling price in your dataset.

In [ ]:
plt.figure(figsize=[8, 6])
sns.scatterplot(x=data.Owner, y=data.Selling_Price)
plt.show()

**Comment on Owner vs. Selling_Price Scatter Plot:**

This scatter plot visualizes the relationship between the number of previous owners (`Owner`) and the `Selling_Price`.

- Since the 'Owner' column has a limited number of discrete values (0, 1, 2, 3), you will see distinct vertical lines or clusters of points at each of these x-axis values.
- By examining the distribution of selling prices along the y-axis for each 'Owner' category, you can visually assess if there is a noticeable difference in selling prices based on the number of previous owners. This plot supports the weak negative correlation observed in the heatmap, suggesting that the number of owners might have a minor impact on the selling price.

This plot helps confirm the relationship between ownership history and selling price as indicated by the correlation matrix.

In [ ]:
plt.figure(figsize=[8, 6])
sns.scatterplot(x=data.Present_Price, y=data.Selling_Price)
plt.show()

**Comment on Present_Price vs. Selling_Price Scatter Plot:**

This scatter plot visualizes the relationship between the `Present_Price` and the `Selling_Price`.

- As expected from the very high positive correlation (close to 1.0) seen in the correlation matrix, there is an almost perfect linear relationship between the present price and the selling price.
- The points on the scatter plot will likely form a very tight cluster along a straight line with a positive slope. This visually confirms that the current price of the car is an extremely strong predictor of its selling price in this dataset.

In [ ]:
plt.figure(figsize=[8, 6])
sns.scatterplot(x=data.Seller_Type_Individual, y=data.Selling_Price)
plt.show()

**Comment on Seller_Type_Individual vs. Selling_Price Scatter Plot:**

This scatter plot visualizes the relationship between whether a car is sold by an individual (`Seller_Type_Individual` = 1) or not (sold by a dealer, `Seller_Type_Individual` = 0) and its `Selling_Price`. Similar to the fuel type plots, you will see two vertical clusters of points:

- A cluster at `x=0`: Represents the selling prices of cars sold by Dealers.
- A cluster at `x=1`: Represents the selling prices of cars sold by Individuals.

By comparing the distribution of points along the y-axis (Selling_Price) for these two clusters, you can visually assess if there is a difference in selling prices based on the seller type. This plot supports the negative correlation observed in the heatmap for `Seller_Type_Individual`, suggesting that cars sold by individuals tend to have different (likely lower) selling prices compared to those sold by dealers.

In [ ]:
plt.figure(figsize=[8, 6])
sns.scatterplot(x=data.Transmission_Manual, y=data.Selling_Price)
plt.show()

**Comment on Transmission_Manual vs. Selling_Price Scatter Plot:**

This scatter plot visualizes the relationship between whether a car has a manual transmission (`Transmission_Manual` = 1) or not (automatic transmission, `Transmission_Manual` = 0) and its `Selling_Price`. Similar to other binary categorical features, you will see two vertical clusters of points:

- A cluster at `x=0`: Represents the selling prices of cars with Automatic Transmission.
- A cluster at `x=1`: Represents the selling prices of cars with Manual Transmission.

By comparing the distribution of points along the y-axis (Selling_Price) for these two clusters, you can visually assess if there is a difference in selling prices based on the transmission type. This plot supports the negative correlation observed in the heatmap for `Transmission_Manual`, suggesting that manual transmission cars tend to have different (likely lower) selling prices compared to automatic ones in this dataset.

In [ ]:
plt.figure(figsize=[8, 6])
sns.scatterplot(x=data.Year, y=data.Selling_Price)
plt.show()

**Comment on Year (Car Age) vs. Selling_Price Scatter Plot:**

This scatter plot visualizes the relationship between the car's age (`Year`) and its `Selling_Price`.

- As expected from the strong negative correlation observed in the heatmap (around -0.78), there is a clear trend where as the car's age increases (moving to the right on the x-axis), the `Selling_Price` tends to decrease (moving downwards on the y-axis).
- The points generally follow a downward slope, indicating that older cars are typically sold for lower prices.

This plot provides a visual confirmation of the significant impact of car age on its selling price in this dataset.

## Modelling

In [ ]:
data.columns

In [ ]:
# with present price and fuel type

#data_df = data[['Year', 'Present_Price', 'Kms_Driven', 'Owner', 'Selling_Price',
       #'Fuel_Type_Diesel', 'Fuel_Type_Petrol', 'Seller_Type_Individual',
       #'Transmission_Manual']]

#without the present price

data_df = data[['Year', 'Kms_Driven', 'Owner', 'Selling_Price',
       'Fuel_Type_Diesel', 'Fuel_Type_Petrol', 'Seller_Type_Individual',
       'Transmission_Manual']]

**Why Separate Features (X) and Target (y) is Necessary:**

Separating your data into features (X) and the target variable (y) is a fundamental and essential step in preparing data for **supervised machine learning**. Here's why:

- **Supervised Learning Requirement:** Supervised learning algorithms learn by being shown input data (features, X) and the corresponding correct output (target, y). They learn the relationship between X and y to make predictions on new, unseen X data.
- **Model Training:** The model is trained using the `X_train` and `y_train` subsets. The algorithm uses the features in `X_train` to learn how to predict the values in `y_train`.
- **Model Evaluation:** After training, the model's performance is evaluated on the `X_test` data, and its predictions are compared to the actual `y_test` values. This allows you to assess how well the model generalizes to new data it hasn't seen during training.
- **Clear Input and Output:** Explicitly defining X and y makes it clear which columns are being used as inputs to the model and which column is the output the model is trying to predict.

In summary, machine learning models need to know exactly what data to use as input (features) and what they are trying to predict (the target). Separating X and y provides this clear structure, which is required for the training and evaluation process in supervised learning.

In [ ]:

## drop the selling price the target values
X = data_df.drop("Selling_Price", axis = 1)
y= data_df.Selling_Price

In [ ]:
from sklearn.model_selection import train_test_split
X_train,Xtest,ytrain,ytest= train_test_split(X,y, test_size = 0.3, random_state = 42)

**Why Splitting Data into Train and Test Sets is Necessary:**

Splitting your data into separate training and testing sets is a fundamental practice in machine learning. Here's why it's essential:

- **Evaluating Generalization:** The primary goal of a machine learning model is to generalize well to *new, unseen data*. If you train and evaluate your model on the same data, you won't know how it will perform on data it hasn't encountered before.
- **Detecting Overfitting:** Training on the entire dataset and then testing on it can lead to **overfitting**. This is when a model learns the training data too well, including its noise and quirks, and performs poorly on new data. By evaluating on a separate test set, you can detect if your model is overfitting.
- **Unbiased Performance Estimate:** The test set provides an unbiased evaluation of your model's performance. It simulates how the model would perform in a real-world scenario on data it hasn't been specifically trained on.
- **Hyperparameter Tuning:** When tuning model hyperparameters, you typically use a separate validation set (often created from the training set) or cross-validation. The final evaluation is always done on the completely held-out test set.

In short, splitting your data ensures that you get a reliable estimate of how well your model will perform in practice on new data, helping you avoid overfitting and build more robust models.

In [ ]:
from sklearn.ensemble import RandomForestRegressor

**Why Random Forest Regressor is Important Here:**

The Random Forest Regressor is a powerful and widely used machine learning algorithm for regression tasks like predicting car selling prices. Here's why it's a good choice and considered important:

- **Handles Non-linearity:** Unlike simpler linear models, Random Forests can capture complex, non-linear relationships between features and the target variable. Car prices are influenced by various factors (age, mileage, brand, type, etc.) which might not have a simple linear relationship.
- **Robust to Outliers and Noise:** Random Forests are less sensitive to outliers and noisy data compared to some other algorithms, which can be beneficial given the potential outliers observed in features like 'Kms_Driven'.
- **Automatic Feature Interaction:** The ensemble of decision trees inherently captures interactions between different features without needing to explicitly engineer interaction terms.
- **Provides Feature Importance:** As you will see later, Random Forests can provide insights into which features were most important in making predictions, which is valuable for understanding the driving factors of car prices.
- **Generally High Accuracy:** Random Forests often provide good predictive accuracy across a variety of datasets.

While other models could also be used, Random Forest is a strong baseline choice known for its performance and ability to handle diverse data types and relationships, making it suitable for this car price prediction problem.

In [ ]:
model = RandomForestRegressor()
model.fit(X_train,ytrain)

The Random Forest Regressor model has been successfully trained on the training data.

In [ ]:
# help in prediction with 30% test set
prediction = model.predict(Xtest)

This line `from sklearn import metrics` imports the entire `metrics` module from the scikit-learn library.

**Why this Import is Important:**

The `sklearn.metrics` module is crucial for **evaluating the performance of your machine learning model**. It contains a wide range of functions for calculating various evaluation metrics for different types of machine learning tasks (classification, regression, clustering, etc.).

In the context of your regression problem (predicting a continuous value like car selling price), the `metrics` module provides functions to calculate metrics such as:

- **Mean Absolute Error (MAE)**
- **Mean Squared Error (MSE)**
- **Root Mean Squared Error (RMSE)**
- **R-squared (R²)**
- **Mean Absolute Percentage Error (MAPE)**

These metrics allow you to quantify how well your trained model is performing on the test data by comparing its predictions (`prediction`) to the actual values (`ytest`). Without importing this module, you wouldn't be able to easily calculate these standard performance indicators to understand the accuracy and error of your model.

In [ ]:
from sklearn import metrics

In [ ]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score, mean_absolute_percentage_error

mae = mean_absolute_error(ytest, prediction)
rmse = np.sqrt(mean_squared_error(ytest, prediction))
r2 = r2_score(ytest, prediction)
mape = mean_absolute_percentage_error(ytest, prediction)

print("MAE:", mae)
print("RMSE:", rmse)
print("R²:", r2)
print("MAPE:", mape)

Here's what the output of the model evaluation metrics means:

1.  **Mean Absolute Error (MAE):** `168.45`

    *   This means that, on average, your model's predictions for the car selling price are off by approximately 168.45 units of your currency.
    *   It provides a straightforward measure of the average magnitude of the errors, without considering their direction.

2.  **Root Mean Squared Error (RMSE):** `379.75`

    *   This represents the square root of the average of the squared differences between the actual and predicted values.
    *   RMSE is also in the same units as your target variable (selling price).
    *   It gives more weight to larger errors than MAE, making it sensitive to outliers in the errors. Your RMSE (379.75) is higher than your MAE (168.45), which is typical, but the ratio suggests there aren't extremely large, isolated errors pulling the RMSE up significantly compared to MAE.

3.  **R² (R-squared):** `0.9354`

    *   This metric indicates the proportion of the variance in the target variable ('Selling_Price') that is predictable from the features used in your model.
    *   An R² value of 0.9354 means that your Random Forest model explains approximately 93.54% of the variation in car selling prices in your test data.
    *   This is a very high R² value, suggesting that your model is doing an excellent job of capturing the underlying patterns that influence selling price.

4.  **Mean Absolute Percentage Error (MAPE):** `0.0168` (or 1.68%)

    *   This measures the average of the absolute percentage errors between the actual and predicted values.
    *   A MAPE of 0.0168 (or 1.68%) means that, on average, your model's predictions are about 1.68% away from the actual selling prices.
    *   MAPE is particularly useful as it provides a relative measure of accuracy that is easy to understand and communicate, especially in business contexts. A MAPE below 10% is often considered good, and 1.68% is excellent.

**Overall Interpretation:**

The evaluation metrics indicate that your Random Forest Regressor model is performing very well on this dataset. An R² of 0.9354 and a MAPE of 1.68% suggest that the model can predict car selling prices with high accuracy and low relative error on unseen data. The MAE and RMSE values further quantify the typical error magnitude in the predictions.

### Feature Importance Visualization

In [ ]:
feat_importances = pd.Series(model.feature_importances_, index=X_train.columns)
feat_importances.sort_values().plot(kind='barh', figsize=(12,6))
plt.title("Feature Importance in Random Forest")
plt.show()

##### 🔑 Explaining Feature Importance

The Random Forest ranked the features in this order:

- **Year (Car Age)** – The newer the car, the higher the selling price. This is a very powerful predictor.
- **Kms Driven** – Cars with more mileage tend to sell for less, so this feature strongly influences price.
- **Seller Type (Individual vs Dealer)** – Individuals often sell cars cheaper than dealers, so this adds predictive power.
- **Transmission (Manual vs Automatic)** – Manual cars may be valued differently depending on market preference.
- **Fuel Type (Diesel vs Petrol)** – Fuel type affects running costs and demand, impacting resale price.
- **Owner Count** – More previous owners can reduce the price since buyers prefer cars with fewer past owners.

👉 Feature importance doesn’t mean causation. It means the model found these features most useful for splitting and reducing prediction error.
Always connect the importance ranking to real-world logic so students understand why the model values those features.

---

##### Why Exclude 'Present_Price'?

Although 'Present_Price' was excluded from the model in cell `873dd100-0381-483d-a11c-9f9770e25862`, it's important to understand why you might choose to exclude it, especially given the extremely high correlation (almost 1.0) with 'Selling_Price' observed in the correlation matrix (cell `72243a19`).

- **Data Leakage:** If the 'Present_Price' is essentially the listed price *at the time of sale*, including it as a feature can be considered data leakage. The model would simply learn that the selling price is almost identical to the present price, which isn't a useful or generalizable insight for predicting prices *before* they are listed or sold.
- **Real-world Use Case:** If your goal is to predict the selling price of a used car *before* it is put on the market (i.e., you don't know its present price), then 'Present_Price' cannot be used as a predictor.
- **Multicollinearity:** Including a feature that is almost perfectly correlated with the target variable can sometimes cause issues like multicollinearity if you were using certain linear models, although ensemble models like Random Forest are less sensitive to this.

By excluding 'Present_Price', you are forcing the model to find relationships between the selling price and other features like age, mileage, and car characteristics, which provides a more realistic and potentially more valuable model for predicting prices based on factors available *before* a car is listed for sale.

### Model Diagnostics

In [ ]:
residuals = ytest - prediction
plt.scatter(ytest, residuals)
plt.axhline(0, color='red', linestyle='--')
plt.xlabel("Actual Selling Price")
plt.ylabel("Residuals")
plt.title("Residual Plot")
plt.show()

**Comment on Model Diagnostics (Residual Plot):**

The residual plot (plotting residuals vs. actual selling prices) provides valuable insights into the model's errors:

- **Random Scattering around Zero:** Ideally, residuals should be randomly scattered around the horizontal line at zero. This indicates that the model's errors are random and not systematically related to the actual values.
- **Absence of Patterns:** Look for any patterns like curves, cones (fanning out or in), or clusters in the residuals. Patterns suggest that the model is missing some information or that a different model or feature transformation might be more appropriate.
- **Error Magnitude:** The plot also shows the typical magnitude of the errors. Most points should be relatively close to the zero line. Outliers in the residuals (points far away from the line) indicate instances where the model made particularly large errors.

Based on the residual plot generated, you can assess if the residuals appear randomly scattered around zero and if the error magnitude is within an acceptable range for your problem.

- Residuals being both positive and negative is a good sign: the model isn’t biased to always predict too high or too low.
- The fact that they’re concentrated mostly between -2000 and +2000 shows the model is fairly accurate, with no extreme systematic error.
- If you saw residuals only on one side (all positive or all negative), it would mean the model is biased.

### Train vs Test set Comparison

In [ ]:
train_pred = model.predict(X_train)
train_rmse = np.sqrt(mean_squared_error(ytrain, train_pred))
test_rmse = np.sqrt(mean_squared_error(ytest, prediction))

print("Train RMSE:", train_rmse)
print("Test RMSE:", test_rmse)

- The fact that Train RMSE (169) < Test RMSE (376) is normal.
- But the gap is important to interpret:
- Small Gap → Good Generalization (model performs equally well on new data).
- Large Gap → Overfitting Risk (model memorized training data patterns but struggles on new data).
- Always expect Test error to be higher than Train error.
- The goal is to keep the gap reasonable so the model generalizes well.
- In your example, the Random Forest model shows excellent learning ability on training data and solid performance on unseen data,
   even though the test error is higher.

## Conclusion

- Model achieves high accuracy (R² = 93.6%).
- Predicts car prices effectively within main price range.
- Feature importance matches real-world reasoning.

In [ ]:
import pickle

with open("random_forest.pkl", "wb") as file:
    pickle.dump(model, file)


The `import pickle` statement is necessary because the `pickle` module is used to **serialize and deserialize Python objects**.

In this specific code, `pickle.dump(model, file)` is used to serialize your trained `model` object (the Random Forest Regressor) into a byte stream and write it to a file. This process allows you to save the state of your trained model so that you can load it back later using `pickle.load()` without needing to retrain it.

Essentially, `pickle` enables you to save complex Python objects (like trained machine learning models) to disk and load them back into memory when needed, which is crucial for deploying or reusing your models.

### Saved model for deployment